packages/task/src/browser/task-configurations.ts

```diff
@@ -0,0 +1,259 @@
+     * @returns an array of strings, each one being a task label
+     */
+    taskConfigurationChanged: (event: string[]) => void;
+}
+
+/**
+ * Watches potentially two tasks.json files: one in the user's home directory (in $home/.theia/ ),
+ * and  another one in the current workspace ( in $workspace/.theia/ )
+ */
+@injectable()
+export class TaskConfigurations implements Disposable {
+
+    /* Task client to send notifications to.  */
+    protected readonly toDispose = new DisposableCollection();
+
+    // Map where all the parsed tasks, from tasks.json files, are saved
+    protected tasksMap = new Map<string, ITaskOptions>();
+
+    // the tasks.json config file(s) we are watching
+    protected watchedConfigFileUris: string[] = [];
+
+    /** last directory element under which we look for task config */
+    protected readonly TASKFILEPATH = '.theia';
+    /** task configuration file name */
+    protected readonly TASKFILE = 'tasks.json';
+
+    protected client: TaskConfigurationClient | undefined = undefined;
+
+    constructor(
+        @inject(ILogger) protected readonly logger: ILogger,
+        @inject(FileSystemWatcherServer) protected readonly watcherServer: FileSystemWatcherServer,
+        @inject(FileSystem) protected readonly fileSystem: FileSystem
+    ) {
+        this.toDispose.push(watcherServer);
+        watcherServer.setClient({
+            onDidFilesChanged: changes =>
+                this.onDidTaskFileChange(changes)
+                    .then(() => {
+                        this.getTaskLabels().then(tasks => {
+                            if (this.client !== undefined) {
+                                this.client.taskConfigurationChanged(tasks);
+                            }
+                        });
+                    })
+                    .catch(err => {
+                        this.logger.info(err);
+                    })
+        });
+
+        this.toDispose.push(Disposable.create(() => {
+            this.tasksMap.clear();
+            this.client = undefined;
+        }));
+    }
+
+    init(client: TaskConfigurationClient) {
+        this.client = client;
+    }
+
+    dispose() {
+        this.toDispose.dispose();
+    }
+
+    watchWorkspaceConfiguration(wsRootUri: string): Promise<void> {
+        return this.watchConfigurationFile(wsRootUri);
+    }
+
+    protected watchConfigurationFile(rootDir: string): Promise<void> {
+        return new Promise(resolve => {
+            let configFile: string;
+            this.getConfigFile(rootDir)
+                .then(uri => {
+                    // start watching that config file
+                    configFile = uri;
+                    if (this.watchedConfigFileUris.indexOf(configFile) === -1) {
+                        this.watchedConfigFileUris.push(configFile);
+                        this.watchConfigFile(configFile);
+                    }
+
+                })
+                .then(() => {
```

Comment:
Oh, so IIUC, in the case we can read a config file but can't watch it, it would be better to successfully read the task configurations from it, even if we are unable to watch the config file further? 

I have tried, without success, to trigger this potential issue, by setting "fs.inotify.max_user_watches" to a really low value, even zero, and still the watches seem to work.

Reply:
>Oh, so IIUC, in the case we can read a config file but can't watch it, it would be better to successfully read the task configurations from it, even if we are unable to watch the config file further?

Yes

>I have tried, without success, to trigger this potential issue, by setting "fs.inotify.max_user_watches" to a really low value, even zero, and still the watches seem to work.

That's because there's possibly a  bug see : #938  Actually it seems like it falls back to polling in that case so it may still work.

Maybe it's not so important in this case.  We can improve in another PR if it's an issue.